docs/src/examples/neural_sde.md

```diff
@@ -0,0 +1,206 @@
+`neural_dmsde` which can be used for the multiplicative noise case (consult the
+layers API documentation, or [this full example using the layer
+function](https://github.com/MikeInnes/zygote-paper/blob/master/neural_sde/neural_sde.jl)).
+
+However, since there are far too many possible combinations for the API to
+support, in many cases you will want to performantly define neural differential
+equations for non-ODE systems from scratch. For these systems, it is generally
+best to use `TrackerAdjoint` with non-mutating (out-of-place) forms. For
+example, the following defines a neural SDE with neural networks for both the
+drift and diffusion terms:
+
+```julia
+dudt(u, p, t) = model(u)
+g(u, p, t) = model2(u)
+prob = SDEProblem(dudt, g, x, tspan, nothing)
+```
+
+where `model` and `model2` are different neural networks. The same can apply to
+a neural delay differential equation. Its out-of-place formulation is
+`f(u,h,p,t)`. Thus for example, if we want to define a neural delay differential
+equation which uses the history value at `p.tau` in the past, we can define:
+
+```julia
+dudt!(u, h, p, t) = model([u; h(t - p.tau)])
+prob = DDEProblem(dudt_, u0, h, tspan, nothing)
+```
+
+
+First let's build training data from the same example as the neural ODE:
+
+```@example nsde
+using Plots, Statistics
+using Lux, Optimization, OptimizationFlux, DiffEqFlux, StochasticDiffEq, SciMLBase.EnsembleAnalysis, Random
+
+rng = Random.default_rng()
+u0 = Float32[2.; 0.]
+datasize = 30
+tspan = (0.0f0, 1.0f0)
+tsteps = range(tspan[1], tspan[2], length = datasize)
+```
+
+```@example nsde
+function trueSDEfunc(du, u, p, t)
+    true_A = [-0.1 2.0; -2.0 -0.1]
+    du .= ((u.^3)'true_A)'
+end
+
+mp = Float32[0.2, 0.2]
+function true_noise_func(du, u, p, t)
+    du .= mp.*u
+end
+
+prob_truesde = SDEProblem(trueSDEfunc, true_noise_func, u0, tspan)
+```
+
+For our dataset we will use DifferentialEquations.jl's [parallel ensemble
+interface](http://docs.juliadiffeq.org/dev/features/ensemble.html) to generate
+data from the average of 10,000 runs of the SDE:
+
+```@example nsde
+# Take a typical sample from the mean
+ensemble_prob = EnsembleProblem(prob_truesde)
+ensemble_sol = solve(ensemble_prob, SOSRI(), trajectories = 10000)
+ensemble_sum = EnsembleSummary(ensemble_sol)
+
+sde_data, sde_data_vars = Array.(timeseries_point_meanvar(ensemble_sol, tsteps))
+```
+
+Now we build a neural SDE. For simplicity we will use the `NeuralDSDE`
+neural SDE with diagonal noise layer function:
+
+```@example nsde
+drift_dudt = Lux.Chain(ActivationFunction(x -> x.^3),
+                       Lux.Dense(2, 50, tanh),
+                       Lux.Dense(50, 2))
+p1, st1 = Lux.setup(rng, drift_dudt)
+
+diffusion_dudt = Lux.Chain(Lux.Dense(2, 2))
+p2, st2 = Lux.setup(rng, diffusion_dudt)
+
+p1 = Lux.ComponentArray(p1)
+p2 = Lux.ComponentArray(p2)
+#Component Arrays doesn't provide a name to the first ComponentVector, only subsequent ones get a name for dereferencing
+p = [p1, p2]
```

Comment:
I will write a manual section later tonight in Lux itself, but this not quite the right interface. The "correct" thing to do would be 
```suggestion
p = Lux.ComponentArray((p1 = p1, p2 = p2))
```
Then index the parameters with `.p1` and `.p2`. This makes it consistent with the Lux ContainerLayer Interface.

The same thing holds for the states

Reply:
Would it be alright to make the user follow a strict naming scheme of p1 and p2, just a small concern.

Reply:
So the user should not be responsible for constructing this in general. Lux.setup on NeuralDSDE should return the parameters in this format

Reply:
The convention I follow in Lux is to document what the parameter and state names are http://lux.csail.mit.edu/stable/devdocs/style_guide/#documentation-for-layers, and construct the structure in `Lux.setup`. After that, if the user messes up then nothing can be done but redirect them to the documentation

Reply:
Okay, lets finalize this for all the Lux constructors in DiffEqFlux then

Reply:
Yes, the user shouldn't construct this in general. This is just a start from a copy paste of broken code so don't take it as a correct way to do things for this. As mentioned in https://github.com/SciML/DiffEqFlux.jl/issues/727, these should be an `AbstractExplicitLayer`, which means they should do things exactly like `Dense`. They should have one state, take in a state, and return a state. They should take in a neural network definition and give you back a state from `setup`. Basically, it should act exactly like Dense does, and be able to perfectly swap in without any other code changes, and if not it's wrong. The only thing that should be different is the constructor for the layer.